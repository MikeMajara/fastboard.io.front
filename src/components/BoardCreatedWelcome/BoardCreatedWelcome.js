import React, {
  useCallback, useEffect, useMemo, useRef, useState,
} from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import { escapeKeyCode } from '#constants';
import { toClipboard } from '#utils';
import { useKey } from '#hooks';
import routes from '#routes';

import Button from '../Button';
import {
  SWrapper, SHeader, SContent, SLink, SCopyIcon, SFooter,
} from './styled';

const copiedMessageTimeout = 2000;

// state -> i18n text key
const copySteps = Object.freeze({
  none      : 'copy',
  copied    : 'copied',
  copyError : 'copyError',
});

const BoardCreatedWelcome = ({ boardId, onClose }) => {
  const { t } = useTranslation('board');
  const timeoutRef = useRef();
  const [copyStep, setCopyStep] = useState(copySteps.none);

  useKey(escapeKeyCode, onClose);

  const boardUrl = useMemo(
    () => `${process.env.REACT_APP_BASE_URL}${routes.board.replace(':id', boardId)}`,
    [boardId],
  );

  const copyAction = useCallback(
    () => toClipboard(boardUrl)
      .then(() => setCopyStep(copySteps.copied))
      .catch(() => setCopyStep(copySteps.copyError))
      .finally(() => {
        if (timeoutRef.current) {
          clearTimeout(timeoutRef.current);
        }

        timeoutRef.current = setTimeout(
          () => setCopyStep(copySteps.none),
          copiedMessageTimeout,
        );
      }),
    [boardUrl],
  );

  // Clear timeout if it was still active
  useEffect(
    () => () => {
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }
    },
    [timeoutRef],
  );

  return (
    <SWrapper>
      <SHeader>
        {t('welcome.created.title', { boardId })}
      </SHeader>
      <SContent>
        <div>
          {t('welcome.created.about')}
        </div>
        <SLink id="link">
          {boardUrl}
        </SLink>
        <Button type="secondary" fullWidth onClick={copyAction}>
          <SCopyIcon />
          {t(`welcome.created.${copyStep}`)}
        </Button>
      </SContent>
      <SFooter>
        <Button type="secondary" onClick={onClose}>
          {t('welcome.created.accept')}
        </Button>
      </SFooter>
    </SWrapper>
  );
};

BoardCreatedWelcome.propTypes = {
  boardId : PropTypes.string.isRequired,
  onClose : PropTypes.func.isRequired,
};

export default BoardCreatedWelcome;
