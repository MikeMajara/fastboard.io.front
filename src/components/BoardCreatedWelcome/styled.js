import styled, { css, down } from '@xstyled/styled-components';
import { getFontSize } from '@xstyled/system';
import { Copy as CopyIcon } from '@styled-icons/fa-solid/Copy';

import Card from '../Card';

export const SCard = styled(Card)`
  display: flex;
  flex-direction: column;
  justify-content: center;
  ${down('md', css`text-align: center;`)}
`;

export const SHeader = styled.h2`
  margin: 0;
  font-size: xlg;
`;

export const SContent = styled.div`
  margin: default 0;
  div {
    margin-bottom: default;
  }
`;

export const SLink = styled.div`
  color: primary;
  font-style: italic;
`;

export const SCopyIcon = styled(CopyIcon)`
  width: ${getFontSize('default')};
  height: ${getFontSize('default')};
  margin-right: xs;
  color: fg;
`;

export const SFooter = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  ${down('md', css`button { width: 100%; }`)}
`;
