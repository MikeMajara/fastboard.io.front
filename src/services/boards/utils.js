/**
 * Send boards event trough the network.
 *
 * @param {String} eventName Event key name.
 * @param {Object} data Data associated to the event.
 * @param {Function} ask Callback for server event acknowledgement.
 */
function send(eventName, data, ack) {
  this.socket.emit(eventName, data, ack);
}

/**
 * Check if the service is initialized.
 *
 * @return {Boolean} True if the service is initialized; false otherwise.
 */
function isInit() {
  return this.isInit;
}

export {
  isInit,
  send,
};
