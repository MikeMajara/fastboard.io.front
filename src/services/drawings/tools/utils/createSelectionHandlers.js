import { Group, Shape } from 'paper';

import { Log } from '#utils';
import { cursorTypes, selectionColorCode } from '#constants';
import store, { setSelectorCursorHover } from '#store';

import { bounds } from '../selector/constants';

const boundsArray = Object.values(bounds);
const bounds2cursorType = Object.freeze({
  [bounds.topLeft]     : cursorTypes.rotating,
  [bounds.topRight]    : cursorTypes.menu,
  [bounds.bottomLeft]  : cursorTypes.resizingSW,
  [bounds.bottomRight] : cursorTypes.resizingSE,
});

/**
 * Add selection handlers to the selected item.
 *
 * @param {Object} selectedItem Item to be added selection handlers to.
 * @param {Object} selectionLayer Layer that will contain the selection handlers.
 *
 * @returns {Object} Item - group with 4 handler items - from the selection handler layer.
 */
export default function createSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : createSelectionHandlers', { selectedItem });

  // Set selected item to show its selection state
  selectedItem.set({
    selected      : true,
    selectedColor : selectionColorCode,
    onMouseEnter  : () => store.dispatch(setSelectorCursorHover(cursorTypes.dragging)),
    onMouseLeave  : () => store.dispatch(setSelectorCursorHover()),
  });

  // Create bound handlers on the selection layer
  boundsArray.forEach((bound) => {
    selectionLayer.addChild(new Shape.Circle({
      name         : bound,
      center       : selectedItem.internalBounds[bound],
      radius       : 6,
      fillColor    : selectionColorCode,
      onMouseEnter : () => store.dispatch(setSelectorCursorHover(bounds2cursorType[bound])),
      onMouseLeave : () => store.dispatch(setSelectorCursorHover()),
    }));
  });

  // Create a group with the handlers to manipulate all of them at once
  selectionLayer.addChild(
    new Group({
      name     : 'selectionHandlers', // @todo to constants
      children : boundsArray.map(
        (bound) => selectionLayer.children[bound],
      ),
    }),
  );

  // Reproduce the selected item transformations on the handler items group
  selectionLayer.children.selectionHandlers.transform(selectedItem.matrix);

  // Except the scaling of the handler items themselves, which retain the original size
  selectionLayer.children.selectionHandlers.children.forEach((handlerItem) => {
    handlerItem.set({ scaling : 1 });
  });

  return selectionLayer.children.selectionHandlers;
}
