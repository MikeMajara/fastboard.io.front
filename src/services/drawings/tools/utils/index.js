export { default as createSelectionHandlers } from './createSelectionHandlers';
export { default as removeSelectionHandlers } from './removeSelectionHandlers';
export { default as resizeSelectionHandlers } from './resizeSelectionHandlers';
export { default as rotateSelectionHandlers } from './rotateSelectionHandlers';
export { default as translateSelectionHandlers } from './translateSelectionHandlers';
