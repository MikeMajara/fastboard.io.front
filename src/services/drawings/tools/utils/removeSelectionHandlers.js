import { Log } from '#utils';

/**
 * Remove selection handlers.
 *
 * @param {Object} selectedItem Item to be removed selection handlers from.
 * @param {Object} selectionLayer Layer that contains the selection handlers.
 */
export default function removeSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : removeSelectionHandlers');

  selectedItem.set({ selected : false });
  selectionLayer.removeChildren();
}
