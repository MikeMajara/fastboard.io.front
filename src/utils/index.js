export { default as Log } from './logger';
export { default as generateId } from './generateId';
export { default as point2net } from './point2net';
export { default as setPreventUnload } from './setPreventUnload';
export { default as throttle } from './throttle';
export { default as timeoutPromise } from './timeoutPromise';
export { default as triggerDownload } from './triggerDownload';
export { default as triggerUpload } from './triggerUpload';
export { default as validBoardId } from './validBoardId';
