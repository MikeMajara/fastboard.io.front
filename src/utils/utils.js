/**
 * Throttles a function call.
 *
 * @param {Function} callback Function to be throttled.
 * @param {Integer} delay Throttle delay in milliseconds.
 *
 * @return {Function} Function that throttles the execution of the callback.
 */
export const throttle = (callback, delay = 0) => {
  let previousTime = new Date().getTime();

  return function throttledCallback(args) {
    const currentTime = new Date().getTime();

    if ((currentTime - previousTime) >= delay) {
      previousTime = currentTime;
      callback(...args);
    }
  };
};

/**
 * Creates a Promise with a rejection timeout.
 *
 * @param {Function} callback Function for the promise. Receives res and rej functions.
 * @param {Integer} timeout Timeout for the promise to be rejected in milliseconds.
 *
 * @return {Promise} Promise with rejection timeout.
 */
export const timeoutPromise = (callback, timeout) => new Promise(
  (res, rej) => {
    const timeoutId = setTimeout(() => rej(new Error()), timeout);

    callback(
      (...args) => {
        clearTimeout(timeoutId);
        res(...args);
      },
      (...args) => {
        clearTimeout(timeoutId);
        rej(new Error(...args));
      },
    );
  },
);
