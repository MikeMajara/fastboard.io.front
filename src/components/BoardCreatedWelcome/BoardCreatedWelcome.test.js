import React from 'react';
import { act } from 'react-dom/test-utils';
import { fireEvent } from '@testing-library/react';

import { render } from '#devTools/tests/utils';
import { toClipboard } from '#utils';

import BoardCreatedWelcome from './BoardCreatedWelcome';

const boardId = '123456';

jest.mock('#utils', () => ({
  toClipboard : jest.fn(() => new Promise((res) => res())),
}));

describe('Component : BoardCreatedWelcome', () => {
  beforeEach(() => {
    toClipboard.mockClear();
  });

  test('Is visible and reacts to actions', async () => {
    const onCloseFn = jest.fn();
    const { findByText, getByText } = render(
      <BoardCreatedWelcome
        boardId={boardId}
        onClose={onCloseFn}
      />,
    );

    expect(toClipboard).toHaveBeenCalledTimes(0);
    await act(async () => fireEvent.click(await findByText('welcome.created.copy')));
    expect(toClipboard).toHaveBeenCalledTimes(1);
    fireEvent.click(getByText('welcome.created.accept'));
    expect(onCloseFn).toHaveBeenCalledTimes(1);
  });

  test('Snapshot', () => {
    const { asFragment } = render(
      <BoardCreatedWelcome
        boardId={boardId}
        onClose={jest.fn()}
      />,
    );

    const componentRender = asFragment();
    expect(componentRender).toMatchSnapshot();
  });
});
