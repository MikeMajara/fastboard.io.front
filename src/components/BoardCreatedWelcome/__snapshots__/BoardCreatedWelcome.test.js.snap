// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component : BoardCreatedWelcome Snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-qZtCU quiCN"
  >
    <h2
      class="sc-pzMyG hXmaiE"
    >
      welcome.created.title
    </h2>
    <div
      class="sc-pHIBf hfxtqR"
    >
      <div>
        welcome.created.about
      </div>
      <div
        class="sc-pQEbo cNjqCi"
        id="link"
      >
        undefined/board/123456
      </div>
      <button
        class="sc-qQYBZ jntorP"
        type="secondary"
      >
        <svg
          aria-hidden="true"
          class="StyledIconBase-ea9ulj-0 jVnGjM sc-pYA-dN kRBtXC"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 00358.059 0H352v96h96v-6.059a24 24 0 00-7.029-16.97z"
            fill="currentColor"
          />
        </svg>
        welcome.created.copy
      </button>
    </div>
    <div
      class="sc-oUcyK cTibrb"
    >
      <button
        class="sc-qQYBZ hUMZgX"
        type="secondary"
      >
        welcome.created.accept
      </button>
    </div>
  </div>
</DocumentFragment>
`;
